File: Class/Calculater.py
class Cal(object):
    def __init__(self, v1, v2): #언더바 두 개를 양 옆에, self를 안 넣고 그냥 v1 v2 하면 위치가 두 개 바뀌어서 들어감...!
        print(v1, v2) #파이썬의 첫 번째 매개변수(self)는 꼭 정의해줘야 하고, 그 첫 번째 매개변수는 그 instance가 됨.
        self.v1 = v1 #instance 변수로 만들기 위해
        self.v2 = v2
    def add(self):
        return self.v1 + self.v2
    def subtract(self):
        return self.v1 - self.v2

c1 = Cal(10, 10)
print(c1.add())
print(c1.subtract())

c2 = Cal(30, 20)
print(c2.add())
print(c2.subtract())

#instance를 사용하는 코드

File: Mixin/calculater.rb
module CalMultiply #클래스는 믹스인을 못하니 모듈로 해줘야 함
  def multiply()
    return @v1 * @v2
  end
end

module CalDivide #클래스는 믹스인을 못하니 모듈로 해줘야 함
  def divide()
    return @v1 / @v2
  end
end

class Cal #객체 이름은 대문자여야 해
  include CalMultiply, CalDivide #Mixin
  def initialize(v1, v2) #루비가 자동으로 실행하는 초기화 함수. instance가 실행될 때 꼭 실행돼. 뒤에 오는 애는 받아오는 매개변수
    p v1, v2 #initialize는 생성자.
    @v1 = v1 #@가 있는 변수는 instance 변수. 인스턴스에 소속된 모든 method에서 사용 가능.
    @v2 = v2 #@가 있으면, 그 instance 내 모든 method에서 접근 가능!
  end
#@가 없는 변수는 method 안에서만 쓸 수 있는 지역변수
#@ 안 달고 내가 initialze에만 name을 초기화한다든지 하면, 이 initialize에서만 쓸 수 있게 되는 거.
  def add()
    return @v1 + @v2 #그냥 v1 v2로 하면 변수 인식 못 해
  end
  def subtract()
    return @v1 - @v2
  end
end

c = Cal.new(100, 10)
p c.add()
p c.multiply()
p c.divide()
